### PR TITLE
EXAMPLES: Add custom buffer size for UCP example.

### LINKED ISSUE #989: Support custom buffer size in ucp_hello_world.c example

### PR DESCRIPTION
**Pull request by guy101**
Fixes #989

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2444/ for details.

File: test/examples/ucp_hello_world.c [hunk 1]
```diff
@@ -71,11 +73,13 @@ static ucp_address_t *peer_addr;
 static size_t local_addr_len;
 static size_t peer_addr_len;
 
-static char *test_str = "Hello UCP World!!!!";
+//static char *test_str = "Hello UCP World!!!!";
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: - no c99 comments
- not leftovers of commented out code

File: test/examples/ucp_hello_world.c [hunk 2]
```diff
@@ -379,8 +384,25 @@ static void barrier(int oob_sock)
     recv(oob_sock, &dummy, sizeof(dummy), 0);
 }
 
+static void generate_random_string(char** str, int size)
+{
+    int i;
+    *str = (char*) calloc(size+1, sizeof(char));
+    char *local_str = *str;
+    printf("errno %d - [%s]", errno, strerror(errno));
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: why is that needed?

File: test/examples/ucp_hello_world.c [hunk 3]
```diff
@@ -379,8 +384,25 @@ static void barrier(int oob_sock)
     recv(oob_sock, &dummy, sizeof(dummy), 0);
 }
 
+static void generate_random_string(char** str, int size)
+{
+    int i;
+    *str = (char*) calloc(size+1, sizeof(char));
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: - check for allocation failure
- where is the string released

File: test/examples/ucp_hello_world.c [hunk 4]
```diff
@@ -379,8 +384,25 @@ static void barrier(int oob_sock)
...
+    char *local_str = *str;
+    printf("errno %d - [%s]", errno, strerror(errno));
+    srand(time(NULL)); // randomize seed
+
+    for( i = 0 ; i < size ; ++i) {
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: code style

File: test/examples/ucp_hello_world.c [hunk 5]
```diff
@@ -379,8 +384,25 @@ static void barrier(int oob_sock)
...
+	char c = 'A' + (rand() % 26);
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: variable 'c' is not needed

File: test/examples/ucp_hello_world.c [hunk 6]
```diff
@@ -530,6 +552,13 @@ int parse_cmd(int argc, char * const argv[], char **server_name)
                 return UCS_ERR_UNSUPPORTED;
             }
             break;
+	case 's':
+	    test_string_length = atoi(optarg);
+	    if (server_port <= 0) {
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: bug

File: test/examples/ucp_hello_world.c [hunk 7]
```diff
@@ -63,6 +64,7 @@ enum ucp_test_mode_t {
 } ucp_test_mode = TEST_MODE_PROBE;
 
 static uint16_t server_port = 13337;
+static uint16_t test_string_length = 16;
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: make it long

File: test/examples/ucp_hello_world.c [hunk 8]
```diff
@@ -71,11 +73,13 @@ static ucp_address_t *peer_addr;
 static size_t local_addr_len;
 static size_t peer_addr_len;
 
-static char *test_str = "Hello UCP World!!!!";
+//static char *test_str = "Hello UCP World!!!!";
+char* test_str;
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: static

**Comment by yosefe** [comment]
@guy101 pls also fix compilation errors

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2445/ for details.

File: test/examples/ucp_hello_world.c (hunk 5 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: It makes the code more readable.

File: test/examples/ucp_hello_world.c (hunk 2 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: it makes the code more readable

**Comment by guy101** [comment]
@yosefe please review again.
thanks!

**Comment by yosefe** [comment]
@guy101 pls fix according to CodeStyle and fix the memory leak

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1786/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1787/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2446/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2447/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1788/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1790/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1789/ for details (Mellanox internal link).

**Comment by guy101** [comment]
@yosefe please review again.
thanks!

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2448/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2449/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1791/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2450/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1792/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2452/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2453/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1794/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1795/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
:+1:

**Comment by yosefe** [comment]
@shssf pls take a look as well

File: test/examples/ucp_hello_world.c [hunk 9]
```diff
@@ -329,6 +332,14 @@ static int run_ucx_server(ucp_worker_h ucp_worker)
         goto err;
     }
 
+    test_str = (char*) calloc(test_string_length+1, sizeof(char));
+    if (!test_str) {
+        printf("unable to allocate memory\n");
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: fprintf(stderr,

Comment (shssf) [inline test/examples/ucp_hello_world.c]: And please do the same for other error messages. Error messages must be print into stderr in this example.

File: test/examples/ucp_hello_world.c [hunk 10]
```diff
@@ -329,6 +332,14 @@ static int run_ucx_server(ucp_worker_h ucp_worker)
...
+        goto err_ep;
+    }
+
+    generate_random_string(test_str, test_string_length);
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: str[size] contains 0 after the call. So, one byte allocated as extra space before.
Perhaps, generate_random_string(test_str, test_string_length + 1);

File: test/examples/ucp_hello_world.c [hunk 11]
```diff
@@ -63,6 +64,7 @@ enum ucp_test_mode_t {
 } ucp_test_mode = TEST_MODE_PROBE;
 
 static uint16_t server_port = 13337;
+static long test_string_length = 16;
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: @yosefe  It might be better to handle this variable as local in main() and pass via function parameters.
It will provide ability to make the loop in main() with different message sizes.

File: test/examples/ucp_hello_world.c [hunk 12]
```diff
@@ -530,6 +553,13 @@ int parse_cmd(int argc, char * const argv[], char **server_name)
                 return UCS_ERR_UNSUPPORTED;
             }
             break;
+        case 's':
+            test_string_length = atoi(optarg);
+            if (test_string_length <= 0) {
+                fprintf(stderr,"Wrong string size %d\n", test_string_length);
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: %ld

File: test/examples/ucp_hello_world.c [hunk 13]
```diff
@@ -530,6 +553,13 @@ int parse_cmd(int argc, char * const argv[], char **server_name)
                 return UCS_ERR_UNSUPPORTED;
             }
             break;
+        case 's':
+            test_string_length = atoi(optarg);
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: atol()

**Comment by yosefe** [comment]
@guy101 pls also fix the commit message according to code style

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1821/ for details (Mellanox internal link).

File: test/examples/ucp_hello_world.c (hunk 12 above)
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: also code style - space after `,`

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1846/ for details (Mellanox internal link).

File: test/examples/ucp_hello_world.c (hunk 11 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: the test string is defined globally - hence the default size for the test string should be defined globally.

**Comment by guy101** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1919/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@guy101 you did not address all comments

**Comment by guy101** [comment]
I disagreed with some of the comments.

File: test/examples/ucp_hello_world.c [hunk 14]
```diff
@@ -379,8 +391,19 @@ static void barrier(int oob_sock)
     recv(oob_sock, &dummy, sizeof(dummy), 0);
 }
 
+static void generate_random_string(char *str, int size)
+{
+    int i;
+    srand(time(NULL)); // randomize seed
+    for( i = 0 ; i < size ; ++i) {
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: Code style. Spaces are in wrong places.
Should be `for (i = 0; i < size; ++i) {`

**Comment by guy101** [comment]
@shssf please review again.

**Comment by shssf** [comment]
@guy101 I see nothing from my comments addressed (fprintf(stderr, atol, ld, code style, extra memory allocation)

**Comment by guy101** [comment]
that's odd.
I'll commit again (used -amend before)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2583/ for details.

**Comment by shamisp** [comment]
probably you have to push -f

**Comment by guy101** [comment]
@shssf please review again.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2586/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1942/ for details (Mellanox internal link).

File: test/examples/ucp_hello_world.c [hunk 15]
```diff
@@ -379,8 +391,19 @@ static void barrier(int oob_sock)
...
+    srand(time(NULL)); // randomize seed
+    for(i = 0; i < size; ++i) {
+        str[i] =  'A' + (rand() % 26);
+    }
+    str[i] = 0;
+}
+
 int main(int argc, char **argv)
 {
+
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: extra line

File: test/examples/ucp_hello_world.c [hunk 16]
```diff
@@ -379,8 +391,19 @@ static void barrier(int oob_sock)
     recv(oob_sock, &dummy, sizeof(dummy), 0);
 }
 
+static void generate_random_string(char *str, int size)
+{
+    int i;
+    srand(time(NULL)); // randomize seed
+    for(i = 0; i < size; ++i) {
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: Code style (space between `for` and `(`)

File: test/examples/ucp_hello_world.c [hunk 17]
```diff
@@ -509,7 +532,7 @@ int parse_cmd(int argc, char * const argv[], char **server_name)
 {
     int c = 0, index = 0;
     opterr = 0;
-    while ((c = getopt (argc, argv, "wfbn:p:h")) != -1) {
+    while ((c = getopt (argc, argv, "wfbn:p:s:h")) != -1) {
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: Code style. No need space after `getopt`

File: test/examples/ucp_hello_world.c [hunk 18]
```diff
@@ -329,6 +332,14 @@ static int run_ucx_server(ucp_worker_h ucp_worker)
         goto err;
     }
 
+    test_str = (char*) calloc(test_string_length + 1, sizeof(char));
```
Comment (shssf) [inline test/examples/ucp_hello_world.c]: The real data length to send will be `test_string_length + 1`. I think it is better to use the same data length as the user requested. If the user wants string length 256, it should be 256 (not 257).
Could you please adjust data size to the user request (remove `+ 1`)?

Comment (shssf) [inline test/examples/ucp_hello_world.c]: @guy101 I meant, the user needs 256 bytes to transfer. It means, we have to provide `255` chars + `0`

File: test/examples/ucp_hello_world.c (hunk 17 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: the extra space was there before. I did not add it.

File: test/examples/ucp_hello_world.c (hunk 15 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: done

File: test/examples/ucp_hello_world.c (hunk 16 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: done

**Comment by guy101** [comment]
@shssf please review again.

File: test/examples/ucp_hello_world.c (hunk 17 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: if it someone's else mistake - it has to be corrected in a different commit. named 'style correction'. not in mine.

ps - please note that there are also several error prints without printing to stderr.

Comment (shssf) [inline test/examples/ucp_hello_world.c]: @guy101 yes, you right, this is the issue for previous developers (including me). Could you please fix this minor issue?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2592/ for details.

**Comment by yosefe** [comment]
@shssf pls fix commit message according to code style

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1951/ for details (Mellanox internal link).

**Comment by guy101** [comment]
@yosefe  - commit message has been changed according to code style

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2597/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1954/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@guy101 the commit message says "Merge pull request #999 from guy101/topic/989"

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/8/ for details.

**Comment by guy101** [comment]
@yosefe should be OK now.
if it is not - tell me what is the correct format.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2616/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/14/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1969/ for details (Mellanox internal link).

File: test/examples/ucp_hello_world.c (hunk 18 above)
Comment (shssf) [inline test/examples/ucp_hello_world.c]: @guy101 I assume the user asks for 256 **bytes**, not the 256 **characters**. In this case, you have to provide 255 **characters** + 1 zero terminator = 256 **bytes**.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2641/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/39/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1996/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2643/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/42/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1998/ for details (Mellanox internal link).

**Comment by guy101** [comment]
bot:mlx:retest

**Comment by guy101** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2669/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/68/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2024/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2025/ for details (Mellanox internal link).

**Comment by shssf** [comment]
@guy101 The commit message remains incorrect. `src/ucs/type/worker_param.h` changes shouldn't be here.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2670/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2671/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/69/ for details.

**Comment by guy101** [comment]
@shssf should finally be OK

**Comment by shssf** [comment]
:+1:

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2026/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2027/ for details (Mellanox internal link).

File: test/examples/ucp_hello_world.c [hunk 19]
```diff
@@ -329,6 +332,14 @@ static int run_ucx_server(ucp_worker_h ucp_worker)
         goto err;
     }
 
+    test_str = (char*) calloc(test_string_length, sizeof(char));
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: test_str is not released in case of error
also, can we avoid allocating test_str and generate the data directly to &msg->data?

File: test/examples/ucp_hello_world.c [hunk 20]
```diff
@@ -530,6 +552,13 @@ int parse_cmd(int argc, char * const argv[], char **server_name)
                 return UCS_ERR_UNSUPPORTED;
             }
             break;
+        case 's':
+            test_string_length = atol(optarg);
+            if (test_string_length <= 0) {
+                fprintf(stderr, "Wrong string size %ld\n", test_string_length);
+                return UCS_ERR_UNSUPPORTED;
+            }	
+	    break;
```
Comment (yosefe) [inline test/examples/ucp_hello_world.c]: indentation

Comment (guy101) [inline test/examples/ucp_hello_world.c]: done

File: test/examples/ucp_hello_world.c (hunk 19 above)
Comment (guy101) [inline test/examples/ucp_hello_world.c]: done

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2672/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/73/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2028/ for details (Mellanox internal link).
